File: backend/rag/fusion.py
"""Rank-based fusion for broad retrieval candidates."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class Candidate:
    card_id: str
    card: dict[str, Any]
    ranks: dict[str, int] = field(default_factory=dict)
    scores: dict[str, float] = field(default_factory=dict)


def reciprocal_rank_fusion(
    source_results: dict[str, list[dict[str, Any]]],
    k: int = 60,
) -> list[dict[str, Any]]:
    candidates: dict[str, Candidate] = {}

    for source_name, rows in source_results.items():
        for rank, row in enumerate(rows, start=1):
            card_id = str(row.get("card_id") or row.get("id") or "")
            if not card_id:
                continue
            candidate = candidates.get(card_id)
            normalized = _normalize_card(row)
            if candidate is None:
                candidate = Candidate(card_id=card_id, card=normalized)
                candidates[card_id] = candidate
            else:
                candidate.card = _merge_card(candidate.card, normalized)
            candidate.ranks[source_name] = min(rank, candidate.ranks.get(source_name, rank))
            if row.get("score") is not None:
                candidate.scores[source_name] = float(row["score"])

    fused = []
    for candidate in candidates.values():
        retrieval_score = sum(1.0 / (k + rank) for rank in candidate.ranks.values())
        row = dict(candidate.card)
        row["card_id"] = candidate.card_id
        row["retrieval_score"] = round(retrieval_score, 6)
        row["source_ranks"] = dict(sorted(candidate.ranks.items()))
        row["source_scores"] = dict(sorted(candidate.scores.items()))
        fused.append(row)

    fused.sort(
        key=lambda row: (
            float(row["retrieval_score"]),
            -min(row["source_ranks"].values()),
        ),
        reverse=True,
    )
    return fused


def _normalize_card(row: dict[str, Any]) -> dict[str, Any]:
    metadata = row.get("metadata") or {}
    return {
        "card_id": row.get("card_id") or row.get("id"),
        "score": row.get("score"),
        "distance": row.get("distance"),
        "section": row.get("section") or row.get("section_name") or metadata.get("section"),
        "tag": row.get("tag") or metadata.get("tag"),
        "card_name": row.get("card_name") or metadata.get("card_name"),
        "argument_name": row.get("argument_name") or metadata.get("argument_name"),
        "citation": row.get("citation") or metadata.get("citation"),
        "author": row.get("author") or metadata.get("author"),
        "year": row.get("year") or metadata.get("year"),
        "document": row.get("document")
        or row.get("document_name")
        or metadata.get("document")
        or metadata.get("document_name"),
        "category": row.get("category") or metadata.get("category"),
        "topical": row.get("topical") if row.get("topical") is not None else metadata.get("topical"),
        "side": row.get("side") or metadata.get("side"),
        "source_path": row.get("source_path") or metadata.get("source_path"),
        "highlights": row.get("highlights") or [],
        "metadata": metadata,
    }


def _merge_card(existing: dict[str, Any], incoming: dict[str, Any]) -> dict[str, Any]:
    merged = dict(existing)
    for key, value in incoming.items():
        if key == "metadata":
            metadata = dict(existing.get("metadata") or {})
            metadata.update(value or {})
            merged[key] = metadata
        elif key == "highlights":
            if not merged.get("highlights") and value:
                merged[key] = value
        elif _is_missing(merged.get(key)) and not _is_missing(value):
            merged[key] = value
    return merged


def _is_missing(value: Any) -> bool:
    return value is None or value == "" or value == []

File: app/backend/hybrid/fusion.cpp
#include "fusion.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace sekret::hybrid {
namespace {

bool missing(const std::string& value) {
    return value.empty();
}

void merge_card(RetrievedCard& existing, const RetrievedCard& incoming) {
    if (missing(existing.section) && !missing(incoming.section)) {
        existing.section = incoming.section;
    }
    if (missing(existing.tag) && !missing(incoming.tag)) {
        existing.tag = incoming.tag;
    }
    if (missing(existing.card_name) && !missing(incoming.card_name)) {
        existing.card_name = incoming.card_name;
    }
    if (missing(existing.argument_name) && !missing(incoming.argument_name)) {
        existing.argument_name = incoming.argument_name;
    }
    if (missing(existing.citation) && !missing(incoming.citation)) {
        existing.citation = incoming.citation;
    }
    if (!existing.author.has_value() && incoming.author.has_value()) {
        existing.author = incoming.author;
    }
    if (!existing.year.has_value() && incoming.year.has_value()) {
        existing.year = incoming.year;
    }
    if (missing(existing.document_name) && !missing(incoming.document_name)) {
        existing.document_name = incoming.document_name;
    }
    if (missing(existing.category) && !missing(incoming.category)) {
        existing.category = incoming.category;
    }
    if (missing(existing.topical) && !missing(incoming.topical)) {
        existing.topical = incoming.topical;
    }
    if (missing(existing.side) && !missing(incoming.side)) {
        existing.side = incoming.side;
    }
    if (missing(existing.source_path) && !missing(incoming.source_path)) {
        existing.source_path = incoming.source_path;
    }
    if (existing.highlights.empty() && !incoming.highlights.empty()) {
        existing.highlights = incoming.highlights;
    }
    if (missing(existing.body_preview) && !missing(incoming.body_preview)) {
        existing.body_preview = incoming.body_preview;
    }
    if (missing(existing.body) && !missing(incoming.body)) {
        existing.body = incoming.body;
    }
}

} // namespace

std::vector<RetrievedCard> reciprocal_rank_fusion(
    const SourceResults& source_results,
    int k
) {
    std::map<std::string, RetrievedCard> candidates;

    for (const auto& [source_name, rows] : source_results) {
        for (std::size_t index = 0; index < rows.size(); ++index) {
            const int rank = static_cast<int>(index + 1);
            const auto& row = rows[index];
            const auto card_id = row.card_id.empty() ? row.id : row.card_id;
            if (card_id.empty()) {
                continue;
            }

            auto [iterator, inserted] = candidates.emplace(card_id, row);
            auto& candidate = iterator->second;
            candidate.card_id = card_id;
            candidate.id = card_id;
            if (!inserted) {
                merge_card(candidate, row);
            }

            auto rank_iterator = candidate.source_ranks.find(source_name);
            if (rank_iterator == candidate.source_ranks.end()) {
                candidate.source_ranks[source_name] = rank;
            } else {
                rank_iterator->second = std::min(rank_iterator->second, rank);
            }
            candidate.source_scores[source_name] = row.score;
        }
    }

    std::vector<RetrievedCard> fused;
    fused.reserve(candidates.size());
    for (auto& [card_id, candidate] : candidates) {
        double retrieval_score = 0.0;
        int best_rank = std::numeric_limits<int>::max();
        for (const auto& [source_name, rank] : candidate.source_ranks) {
            (void)source_name;
            retrieval_score += 1.0 / static_cast<double>(k + rank);
            best_rank = std::min(best_rank, rank);
        }
        candidate.retrieval_score = std::round(retrieval_score * 1000000.0) / 1000000.0;
        candidate.score = candidate.retrieval_score;
        fused.push_back(candidate);
    }

    std::sort(fused.begin(), fused.end(), [](const RetrievedCard& left, const RetrievedCard& right) {
        if (left.retrieval_score != right.retrieval_score) {
            return left.retrieval_score > right.retrieval_score;
        }
        const auto left_rank = std::min_element(
            left.source_ranks.begin(),
            left.source_ranks.end(),
            [](const auto& a, const auto& b) { return a.second < b.second; }
        );
        const auto right_rank = std::min_element(
            right.source_ranks.begin(),
            right.source_ranks.end(),
            [](const auto& a, const auto& b) { return a.second < b.second; }
        );
        const int left_best = left_rank == left.source_ranks.end() ? 0 : left_rank->second;
        const int right_best = right_rank == right.source_ranks.end() ? 0 : right_rank->second;
        return left_best < right_best;
    });
    return fused;
}

} // namespace sekret::hybrid
